### PR TITLE
refactor(pagination): clarify edge cases and throw errors

### PR DESCRIPTION
**Pull request by maxokorokov**
* `ngOnChanges` suspicious calls clarified
* pagination throws errors to avoid confusion when providing non-numeric values for `page`, `pageSize`, `maxSize` and `collectionSize` (as is #898). This will also avoid emitting `NaN` values in `(pageChange)`
* added more tests for edge cases

### CODE REVIEW

File: src/pagination/pagination.ts [hunk 1]
```diff
@@ -224,4 +199,40 @@ export class NgbPagination implements OnChanges {
       this.pageChange.emit(this.page);
     }
   }
+
+  private _updatePages(newPage: number) {
+    // re-calculate new length of pages, should always have at least 1
```
Comment (pkozlowski-opensource) [inline src/pagination/pagination.ts]: Why `1` and not `0`?

File: src/pagination/pagination.spec.ts [hunk 2]
```diff
@@ -440,6 +440,74 @@ describe('ngb-pagination', () => {
...
+
+    describe('invalid inputs', () => {
+
+      ['page', 'collectionSize', 'maxSize', 'pageSize'].forEach((input) => {
+
+        it(`should throw for invalid "${input}" value`, () => {
+          const fixture = createTestComponent(`<ngb-pagination [collectionSize]="collectionSize" 
+            [maxSize]="maxSize" [pageSize]="pageSize" [page]="page"></ngb-pagination>`);
+          fixture.componentInstance[input] = NaN;
+          expect(() => { fixture.detectChanges(); }).toThrow();
```
Comment (pkozlowski-opensource) [inline src/pagination/pagination.spec.ts]: Wouldn't it be good to verify the exact message here? At least for one of inputs.

File: src/pagination/pagination.ts (hunk 1 above)
Comment (maxokorokov) [inline src/pagination/pagination.ts]: I think you always have at least one page in pagination, otherwise it would be only `<` and `>` arrows, which looks strange

**Comment by pkozlowski-opensource** [comment]
@maxokorokov had another look at this PR and I would like to go back to the discussion on 2 items:
* throwing errors on invalid inputs vs. auto-correcting it
* displaying page "1" vs. not displaying anything when the collection size is 0

I'm kind of on the fence when it comes to those 2 items but more leaning towards not throwing and 0 vs. 1. But open to be convinced otherwise :-)

More generally I think I would prefer that we split this PR into 2 parts:
* actual bug fix
* change in behavior

This would make it easier to discuss / review and prioritize.

**Comment by maxokorokov** [comment]
No pb, will open two separate requests
